hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -183,13 +184,20 @@ def __init__(self, element_strategy, shape, dtype, fill, unique):
         self.fill = fill
         self.array_size = int(np.prod(shape))
         self.dtype = dtype
-        self.element_strategy = element_strategy
         self.unique = unique
         self._check_elements = dtype.kind not in ("O", "V")
 
-    def set_element(self, data, result, idx, strategy=None):
-        strategy = strategy or self.element_strategy
-        val = data.draw(strategy)
+        # UniqueSampledListStrategy offers a substantial performance improvement for
+        # unique arrays with few possible elements, e.g. of eight-bit integer types.
+        if unique and fill.is_empty and isinstance(element_strategy, IntegersStrategy):
+            start, end = element_strategy.start, element_strategy.end
+            if None not in (start, end) and (end - start) <= 255:  # pragma: no branch
+                element_strategy = st.sampled_from(
+                    list(range(0, end + 1)) + list(range(0, start, -1))
```

Comment:
This doesn't seem right...

```python
>>> start, end = 2, 4

>>> list(range(0, end + 1)) + list(range(0, start, -1))
[0, 1, 2, 3, 4]
```

and thus you can get behavior like:

```python
>>> @given(arrays(dtype="int8", shape=2, elements=st.integers(2, 4), unique=True))
... def test_unique(x):
...     assert all(2 <= i <= 4 for i in x)
Falsifying example: test_unique(
    x=array([0, 4], dtype=int8),
)

```

Reply:
🤦‍♂️  Yeah, you're right - this is just wrong for any interval that doesn't span zero, and has an off-by-one error for negative parts too.

I've pushed a fix for both, which also generalises this feature by embedding the relevant logic in `lists()` rather than `arrays()` - it's equally useful outside of Numpy, after all!